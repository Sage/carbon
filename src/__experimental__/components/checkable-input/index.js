export { default } from './checkable-input.component';
