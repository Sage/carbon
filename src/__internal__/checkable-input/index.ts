export { default } from "./checkable-input.component";
export type {
  CheckableInputProps,
  CommonCheckableInputProps,
} from "./checkable-input.component";
