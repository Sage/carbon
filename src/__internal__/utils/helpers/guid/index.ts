/**
 * Generates a random guid, useful for creating unique IDs.
 */
export default (): string => {
  function s4(): string {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  }
  return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
};
