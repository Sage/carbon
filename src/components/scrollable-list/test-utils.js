import React from 'react';
import { makeArrayKeys } from '../../__spec_helper__/test-utils';
import ScrollableListItem from './scrollable-list-item.component';

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArrayKeys(opts.num).reduce(listItemReducer(opts), []);
};

export default renderListItems;
