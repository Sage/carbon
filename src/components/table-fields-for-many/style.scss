@import "colors";

.ui-table-fields-for-many {
  border: 1px solid $border-color {
    collapse: separate;
    radius: 4px;
    spacing: 0;
  }
}
