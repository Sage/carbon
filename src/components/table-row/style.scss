@import "colors";

$row__background-color--even: $white;
$row__background-color--odd: #F2F2F2;
$row__background-color--hover: #D2EED1;
$row__border-radius: 4px;
$row__delete-icon-size: 18px;
$row__delete-icon-color--normal: #AFAFAF;
$row__delete-icon-color--hover: $error;

.ui-table-row {
  background-color: $row__background-color--odd;

  // first row
  &:first-child {
    .ui-table-row__td {
      // first td in first row
      &:first-child {
        border-radius: $row__border-radius 0 0 0;
      }

      // last td in first row
      &:last-child {
        border-radius: 0 $row__border-radius 0 0;
      }
    }
  }

  // last row
  &:last-child {
    .ui-table-row__td {
      // first td in last row
      &:first-child {
        border-radius: 0 0 0 $row__border-radius;
      }

      // last td in last row
      &:last-child {
        border-radius: 0 0 $row__border-radius 0;
      }
    }
  }

  // even rows
  &:nth-child(even) {
    background-color: $row__background-color--even;
  }

  // row hover
  &:hover {
    background-color: $row__background-color--hover;

    // delete icon on row hover
    .ui-table-row__delete-icon:before {
      content: "w";
      font-size: $row__delete-icon-size;
    }
  }
}

.ui-table-row__td {
  padding: 4px 2px;
}

.ui-table-row__delete {
  background: none;
  border: none;
  cursor: pointer;
  display: block;
  padding: 0;
  width: $row__delete-icon-size;
}

.ui-table-row__delete-icon {
  &:before {
    color: $row__delete-icon-color--normal;
    font-size: $row__delete-icon-size - 2;
  }

  &:hover:before {
    color: $row__delete-icon-color--hover;
  }
}
