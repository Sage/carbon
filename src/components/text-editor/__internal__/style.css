
.textBold {
  font-weight: bold;
}
.textItalic {
  font-style: italic;
}
