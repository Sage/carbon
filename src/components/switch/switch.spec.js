import React from "react";
import { act } from "react-dom/test-utils";
import { css, ThemeProvider } from "styled-components";
import { mount } from "enzyme";

import Switch from ".";
import CheckableInput from "../../__internal__/checkable-input";
import { StyledCheckableInput } from "../../__internal__/checkable-input/checkable-input.style";
import FieldHelpStyle from "../../__internal__/field-help/field-help.style";
import HiddenCheckableInputStyle from "../../__internal__/checkable-input/hidden-checkable-input.style";
import { StyledLabelContainer } from "../../__internal__/label/label.style";
import StyledSwitchSlider from "./__internal__/switch-slider.style";
import guid from "../../__internal__/utils/helpers/guid";
import {
  assertStyleMatch,
  carbonThemesJestTable,
  mockMatchMedia,
  testStyledSystemMargin,
} from "../../__spec_helper__/test-utils";
import StyledValidationIcon from "../../__internal__/validations/validation-icon.style";
import { baseTheme } from "../../style/themes";
import SwitchSliderPanel from "./__internal__/switch-slider-panel.style";
import SwitchStyle from "./switch.style";
import Label from "../../__internal__/label";
import I18nProvider from "../i18n-provider";
import Tooltip from "../tooltip";
import StyledHelp from "../help/help.style";

jest.mock("../../__internal__/utils/helpers/guid");
guid.mockImplementation(() => "guid-12345");

const getLabel = (wrapper) => wrapper.find(SwitchSliderPanel).text();
const wrappingComponent = (props) => (
  <I18nProvider
    {...props}
    locale={{
      locale: () => "fr-FR",
      switch: {
        on: () => "sur",
        off: () => "de",
      },
    }}
  />
);

describe("Switch", () => {
  describe("Styled System", () => {
    testStyledSystemMargin((props) => <Switch {...props} />);
  });

  describe("uncontrolled behaviour", () => {
    it("sets proper default internal state", () => {
      const wrapper = render({ defaultChecked: true }, mount);
      expect(wrapper.find(CheckableInput).prop("checked")).toBe(true);
    });

    it("changes internal state and passes event to the provided onChange prop when change is triggered", () => {
      const onChangeMock = jest.fn();
      const event = {
        target: {
          checked: true,
          name: "some_name",
        },
      };
      const wrapper = render({ onChange: onChangeMock }, mount);
      act(() => {
        wrapper.find(CheckableInput).prop("onChange")(event);
      });
      expect(onChangeMock).toHaveBeenCalledWith(event);
      wrapper.update();
      expect(wrapper.find(CheckableInput).prop("checked")).toBe(true);
    });
  });

  describe("controlled behaviour", () => {
    it("passes checked value to the CheckableInput", () => {
      const wrapper = render({ checked: true, onChange: jest.fn() }, mount);
      expect(wrapper.find(CheckableInput).prop("checked")).toBe(true);
    });

    it("reacts properly to checked prop change", () => {
      const wrapper = render({ checked: true, onChange: jest.fn() }, mount);
      expect(wrapper.find(CheckableInput).prop("checked")).toBe(true);
      act(() => {
        wrapper.setProps({ checked: false });
      });
      wrapper.update();
      expect(wrapper.find(CheckableInput).prop("checked")).toBe(false);
    });

    it("passes event to the provided onChange prop when change is triggered", () => {
      const onChangeMock = jest.fn();
      const event = {
        target: {
          checked: true,
          name: "some_name",
        },
      };
      const wrapper = render({ checked: false, onChange: onChangeMock }, mount);
      act(() => {
        wrapper.find(CheckableInput).prop("onChange")(event);
      });
      expect(onChangeMock).toHaveBeenCalledWith(event);
    });
  });

  describe("i18n", () => {
    describe("default translation", () => {
      it("has default translation for on", () => {
        const wrapper = render({ checked: true, onChange: jest.fn() }, mount);
        expect(getLabel(wrapper)).toBe("ON");
      });

      it("has default translation for off", () => {
        const wrapper = render({ checked: false, onChange: jest.fn() }, mount);
        expect(getLabel(wrapper)).toBe("OFF");
      });
    });
  });

  describe("base theme", () => {
    describe("when reverse=false", () => {
      describe("default", () => {
        const wrapper = render({ reverse: false });

        it("applies the correct Label styles", () => {
          assertStyleMatch(
            {
              marginTop: "8px",
            },
            wrapper,
            {
              modifier: css`
                ${StyledLabelContainer}
              `,
            }
          );
        });
      });

      describe("and fieldHelpInline=true", () => {
        const wrapper = render({
          reverse: false,
          fieldHelpInline: true,
        });

        it("applies the correct FieldHelp styles", () => {
          assertStyleMatch(
            {
              margin: "0",
              marginTop: "8px",
            },
            wrapper,
            {
              modifier: css`
                ${FieldHelpStyle}
              `,
            }
          );
        });
      });

      describe("and labelInline=true, fieldHelpInline=false", () => {
        const wrapper = render({
          fieldHelpInline: false,
          labelInline: true,
          reverse: false,
        });

        it("applies the correct FieldHelp styles", () => {
          assertStyleMatch(
            {
              marginLeft: "60px",
            },
            wrapper,
            {
              modifier: css`
                ${FieldHelpStyle}
              `,
            }
          );
        });
      });
    });

    describe("when fieldHelpInline=true", () => {
      const wrapper = render({ fieldHelpInline: true });

      it("applies the correct FieldHelp styles", () => {
        assertStyleMatch(
          {
            margin: "0",
          },
          wrapper,
          {
            modifier: css`
              ${FieldHelpStyle}
            `,
          }
        );
      });
    });

    describe("when labelInline=true", () => {
      it("applies the correct Label styles", () => {
        const wrapper = render({ labelInline: true });
        assertStyleMatch(
          {
            marginBottom: "0",
          },
          wrapper,
          {
            modifier: css`
              ${StyledLabelContainer}
            `,
          }
        );
      });

      it("applies the correct FieldHelp styles", () => {
        const wrapper = render({ labelInline: true });
        assertStyleMatch(
          {
            marginTop: "0",
          },
          wrapper,
          {
            modifier: css`
              ${FieldHelpStyle}
            `,
          }
        );
      });

      describe("when adaptiveLabelBreakpoint prop is set", () => {
        describe("when screen bigger than breakpoint", () => {
          beforeEach(() => {
            mockMatchMedia(true);
          });

          it("should pass labelInline to its children", () => {
            const wrapper = render(
              {
                label: "Label",
                labelInline: true,
                adaptiveLabelBreakpoint: 1000,
              },
              mount
            );

            expect(wrapper.find(SwitchStyle).props().labelInline).toEqual(true);
            expect(wrapper.find(CheckableInput).props().labelInline).toEqual(
              true
            );
          });
        });

        describe("when screen smaller than breakpoint", () => {
          beforeEach(() => {
            mockMatchMedia(false);
          });

          it("should pass labelInline as false to its children", () => {
            const wrapper = render(
              {
                label: "Label",
                labelInline: true,
                adaptiveLabelBreakpoint: 1000,
              },
              mount
            );

            expect(wrapper.find(SwitchStyle).props().labelInline).toEqual(
              false
            );
            expect(wrapper.find(CheckableInput).props().labelInline).toEqual(
              false
            );
          });
        });
      });
    });

    describe("when fieldHelpInline=true and labelInline=true", () => {
      const wrapper = render({
        fieldHelpInline: true,
        labelInline: true,
      });

      it("applies the correct CheckableInput styles", () => {
        assertStyleMatch(
          {
            marginLeft: "10px",
          },
          wrapper,
          {
            modifier: css`
              ${StyledCheckableInput}
            `,
          }
        );
      });

      it("applies the correct Label styles", () => {
        assertStyleMatch(
          {
            marginRight: "10px",
          },
          wrapper,
          {
            modifier: css`
              ${StyledLabelContainer}
            `,
          }
        );
      });

      it("applies the correct FieldHelp styles", () => {
        assertStyleMatch(
          {
            marginLeft: "0",
          },
          wrapper,
          {
            modifier: css`
              ${FieldHelpStyle}
            `,
          }
        );
      });
    });

    describe("when size=large", () => {
      describe("default", () => {
        const wrapper = render({ size: "large" });

        const largeSizes = {
          height: "40px",
          width: "78px",
        };

        it("applies the correct CheckableInput styles", () => {
          assertStyleMatch(largeSizes, wrapper, {
            modifier: css`
              ${StyledCheckableInput}
            `,
          });
        });

        it("applies the correct HiddenCheckableInput styles", () => {
          assertStyleMatch(largeSizes, wrapper, {
            modifier: css`
              ${HiddenCheckableInputStyle}
            `,
          });
        });

        it("applies the correct SwitchSlider styles", () => {
          assertStyleMatch(largeSizes, wrapper, {
            modifier: css`
              ${StyledSwitchSlider}
            `,
          });
        });
      });

      describe("and labelInline=true", () => {
        it("applies the correct Label styles", () => {
          const wrapper = render({ size: "large", labelInline: true });

          assertStyleMatch(
            {
              marginTop: "1px",
              paddingTop: "10px",
              paddingBottom: "10px",
            },
            wrapper,
            {
              modifier: css`
                ${StyledLabelContainer}
              `,
            }
          );
        });

        describe("and reverse=false", () => {
          const wrapper = render({
            size: "large",
            labelInline: true,
            reverse: false,
          });

          it("applies the correct FieldHelp styles", () => {
            assertStyleMatch(
              {
                marginLeft: "78px",
              },
              wrapper,
              {
                modifier: css`
                  ${FieldHelpStyle}
                `,
              }
            );
          });
        });
      });
    });
  });

  describe("validations", () => {
    const wrapper = render(
      {
        label: "My Label",
        labelHelp: "Please help me?",
      },
      mount
    );
    const validationTypes = ["error", "warning", "info"];

    beforeEach(() => {
      const props = {
        error: false,
        warning: false,
        info: false,
        validationOnLabel: false,
      };

      wrapper.setProps(props);
    });

    describe.each(validationTypes)("when %s prop passed as string", (type) => {
      it(`displays ${type} icon by the input`, () => {
        wrapper.setProps({
          [type]: "Message",
        });
        expect(
          wrapper
            .find(StyledSwitchSlider)
            .find(StyledValidationIcon)
            .prop("validationType")
        ).toEqual(type);
      });

      it(`displays ${type} icon by the label when validationOnLabel is passed as true`, () => {
        wrapper.setProps({
          [type]: "Message",
          validationOnLabel: true,
        });
        expect(
          wrapper
            .find(StyledLabelContainer)
            .find(StyledValidationIcon)
            .prop("validationType")
        ).toEqual(type);
      });

      it("renders proper validation styles", () => {
        wrapper.setProps({
          [type]: "Message",
        });
        const shadowWidth = type === "error" ? 2 : 1;
        assertStyleMatch(
          {
            boxShadow: `inset 0px 0px 0px ${shadowWidth}px ${baseTheme.colors[type]}`,
          },
          wrapper.find(StyledSwitchSlider)
        );
      });
    });

    describe.each(validationTypes)(
      "when %s prop passed as true boolean",
      (type) => {
        it(`displays ${type} icon by the label when validationOnLabel is passed as true`, () => {
          wrapper.setProps({
            [type]: true,
            validationOnLabel: true,
          });
          expect(wrapper.find(StyledValidationIcon).exists()).toBe(false);
        });

        it("renders proper validation styles", () => {
          wrapper.setProps({
            [type]: true,
          });
          const shadowWidth = type === "error" ? 2 : 1;
          assertStyleMatch(
            {
              boxShadow: `inset 0px 0px 0px ${shadowWidth}px ${baseTheme.colors[type]}`,
            },
            wrapper.find(StyledSwitchSlider)
          );
        });
      }
    );

    it("forces validation icon to be displayed on label when labelInline = true and reverse = false", () => {
      wrapper.setProps({ error: "Error", labelInline: true, reverse: false });

      expect(
        wrapper.find(StyledLabelContainer).find(StyledValidationIcon).exists()
      ).toEqual(true);
      expect(
        wrapper.find(StyledSwitchSlider).find(StyledValidationIcon).exists()
      ).toEqual(false);
    });
  });

  describe("tooltipPosition", () => {
    it("overrides the default placement when icon is on input", () => {
      const { position } = render(
        {
          label: "My Label",
          labelHelp: "Please help me?",
          tooltipPosition: "top",
        },
        mount
      )
        .find(Tooltip)
        .props();

      expect(position).toEqual("top");
    });

    it("overrides the default placement when icon is on label", () => {
      const { position } = render(
        {
          label: "My Label",
          labelHelp: "Please help me?",
          tooltipPosition: "top",
          validationOnLabel: true,
        },
        mount
      )
        .find(Tooltip)
        .props();

      expect(position).toEqual("top");
    });
  });

  describe("helpAriaLabel", () => {
    it("should set the aria-label on the Help component", () => {
      const text = "foo";
      const wrapper = render(
        { label: "foo", labelHelp: text, helpAriaLabel: text },
        mount
      );
      const help = wrapper.find(StyledHelp);

      expect(help.prop("aria-label")).toEqual(text);
    });
  });

  describe.each(carbonThemesJestTable)(
    "when the theme is set to %s",
    (themeName, theme) => {
      describe("default", () => {
        const wrapper = renderWithTheme({}, theme);

        describe("and hover is applied to the element", () => {
          it("then the correct background should be rendered", () => {
            assertStyleMatch(
              {
                backgroundColor: theme.switch.offHover,
              },
              wrapper,
              {
                modifier: css`
                  ${`${HiddenCheckableInputStyle}:not([disabled]):hover + ${StyledSwitchSlider}`}
                `,
              }
            );
          });
        });

        describe("and focus is applied to the element", () => {
          it("then the correct outline should be rendered", () => {
            assertStyleMatch(
              {
                outline: `solid 3px ${theme.colors.focus}`,
              },
              wrapper,
              {
                modifier: css`
                  ${`${HiddenCheckableInputStyle}:not([disabled]):focus + ${StyledSwitchSlider}`}
                `,
              }
            );
          });
        });
      });
    }
  );

  describe("required", () => {
    let wrapper;

    beforeAll(() => {
      wrapper = render({ required: true, label: "required" }, mount);
    });

    it("the required prop is passed to the input", () => {
      const input = wrapper.find("input");
      expect(input.prop("required")).toBe(true);
    });

    it("the isRequired prop is passed to the label", () => {
      const label = wrapper.find(Label);
      expect(label.prop("isRequired")).toBe(true);
    });
  });
});

describe("translation", () => {
  it("can use i18n for on", () => {
    const wrapper = render({ checked: true }, mount, {
      wrappingComponent,
    });

    expect(getLabel(wrapper)).toBe("sur");
  });

  it("can use i18n for off", () => {
    const wrapper = render({ checked: false }, mount, {
      wrappingComponent,
    });
    expect(getLabel(wrapper)).toBe("de");
  });
});

function render(props, renderer = mount, params) {
  return renderer(
    <Switch name="my-switch" value="test" onChange={() => {}} {...props} />,
    params
  );
}

function renderWithTheme(props, theme, renderer = mount) {
  return renderer(
    <ThemeProvider theme={theme}>
      <Switch name="my-switch" value="test" onChange={() => {}} {...props} />
    </ThemeProvider>
  );
}
