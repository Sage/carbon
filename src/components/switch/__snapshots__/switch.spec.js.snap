// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders a Switch component with loading dots, loading className and without the reverse className 1`] = `
<Checkbox
  className="carbon-switch  carbon-switch__loading"
  disabled={true}
  label="My label"
  readOnly={true}
  reverse={false}
>
  <div
    className="carbon-switch__switch carbon-switch__switch__loading"
  >
    <span
      className="carbon-switch__slider"
    />
    <div
      className="carbon-switch__on"
    >
      <div
        className="carbon-loading-dots"
      >
        <div
          className="carbon-loading-dots__bounce carbon-loading-dots__bounce1"
        />
        <div
          className="carbon-loading-dots__bounce carbon-loading-dots__bounce2"
        />
        <div
          className="carbon-loading-dots__bounce carbon-loading-dots__bounce3"
        />
      </div>
    </div>
    <div
      className="carbon-switch__off"
    >
      <div
        className="carbon-loading-dots"
      >
        <div
          className="carbon-loading-dots__bounce carbon-loading-dots__bounce1"
        />
        <div
          className="carbon-loading-dots__bounce carbon-loading-dots__bounce2"
        />
        <div
          className="carbon-loading-dots__bounce carbon-loading-dots__bounce3"
        />
      </div>
    </div>
  </div>
</Checkbox>
`;

exports[`Switch renders a Switch component with reverse className 1`] = `
<Checkbox
  className="carbon-switch  carbon-switch__reverse"
  label="My label"
  reverse={true}
>
  <div
    className="carbon-switch__switch"
  >
    <span
      className="carbon-switch__slider"
    />
    <div
      className="carbon-switch__on"
    >
      <Icon
        bgSize="small"
        disabled={false}
        fontSize="small"
        type="tick"
      />
    </div>
    <div
      className="carbon-switch__off"
    >
      <Icon
        bgSize="small"
        disabled={false}
        fontSize="small"
        type="cross"
      />
    </div>
  </div>
</Checkbox>
`;
